Parcial 1: Fundamentos de Aprendizaje de Maquina

Autor: William Andrés Gómez Roa

06/09/2024

# **ANÁLISIS DE REGRESIÓN / REGULARIZACIÓN**

**Precios de diamantes**

**Objetivo:** Predicción del precio de un diamante con base en atributos físicos del mismo.

Dataset: Diamantes.csv (seleccione 10.000 registros aleatoriamente)

Descripción de los campos:
1. **Indice**: número consecutivo
2. **Precio**: precio en dólares estadounidenses
3. **Quilates**: peso del diamante
4. **Talla**: calidad de la talla (0: regular, 1: buena, 2: ideal, 3: superior, 4: muy buena)
5. **Color**: color del diamante (0: D, 1: E, 2: F, 3: G, 4: H, 5: I, 6: J)
6. **Claridad**: medida de la claridad del diamante (0: I1, 1: IF, 2: SI1, 3: SI2, 4: VS1, 5: VS2, 6: VVS1, 7: VVS2)
7. **Profundidad**: porcentaje de profundidad total
8. **Tabla**: anchura de la parte superior del diamante en relación con el punto más ancho
9. **X**: longitud en mm
10. **Y**: anchura en mm
11. **Z**: profundidad en mm

**Preguntas:**

1. (6) Desarrolle un modelo de regresión lineal múltiple – utilizando el método de stepwise – para estimar el precio de un diamante. Evalúe su desempeño.
2. (12) Desarrolle un modelo de regresión lineal múltiple usando regularización ELASTIC NET. Optimice los parámetros correspondientes. Evalúe su desempeño. Escriba la ecuación final de la función objetivo a minimizar (con los parámetros optimizados). ¿Tiene un mejor desempeño que el modelo sin regularización?

# **SOLUCION**

In [8]:
SEED <- 837

In [5]:
diamante<-read.csv("/content/Diamantes.csv")
str(diamante)

'data.frame':	53940 obs. of  11 variables:
 $ indice     : int  0 1 2 3 4 5 6 7 8 9 ...
 $ quilates   : num  0.23 0.21 0.23 0.29 0.31 0.24 0.24 0.26 0.22 0.23 ...
 $ talla      : int  2 3 1 3 1 4 4 4 0 4 ...
 $ color      : int  1 1 1 5 6 6 5 4 1 4 ...
 $ claridad   : int  3 2 4 5 3 7 6 2 5 4 ...
 $ profundidad: num  61.5 59.8 56.9 62.4 63.3 62.8 62.3 61.9 65.1 59.4 ...
 $ tabla      : num  55 61 65 58 58 57 57 55 61 61 ...
 $ precio     : int  326 326 327 334 335 336 336 337 337 338 ...
 $ x          : num  3.95 3.89 4.05 4.2 4.34 3.94 3.95 4.07 3.87 4 ...
 $ y          : num  3.98 3.84 4.07 4.23 4.35 3.96 3.98 4.11 3.78 4.05 ...
 $ z          : num  2.43 2.31 2.31 2.63 2.75 2.48 2.47 2.53 2.49 2.39 ...


In [6]:
diamante <-diamante[,-1]
str(diamante)

'data.frame':	53940 obs. of  10 variables:
 $ quilates   : num  0.23 0.21 0.23 0.29 0.31 0.24 0.24 0.26 0.22 0.23 ...
 $ talla      : int  2 3 1 3 1 4 4 4 0 4 ...
 $ color      : int  1 1 1 5 6 6 5 4 1 4 ...
 $ claridad   : int  3 2 4 5 3 7 6 2 5 4 ...
 $ profundidad: num  61.5 59.8 56.9 62.4 63.3 62.8 62.3 61.9 65.1 59.4 ...
 $ tabla      : num  55 61 65 58 58 57 57 55 61 61 ...
 $ precio     : int  326 326 327 334 335 336 336 337 337 338 ...
 $ x          : num  3.95 3.89 4.05 4.2 4.34 3.94 3.95 4.07 3.87 4 ...
 $ y          : num  3.98 3.84 4.07 4.23 4.35 3.96 3.98 4.11 3.78 4.05 ...
 $ z          : num  2.43 2.31 2.31 2.63 2.75 2.48 2.47 2.53 2.49 2.39 ...


In [9]:
set.seed(SEED)
sample <- sample.int(nrow(diamante), floor(.75*nrow(diamante)))
diamante.train <- diamante[sample, ]
diamante.test <- diamante[-sample, ]

### **MODELO : REGRESION LINEAL MULTIPLE**

In [11]:
modelo_simple<-lm(precio~.,data=diamante.train)
summary(modelo_simple)


Call:
lm(formula = precio ~ ., data = diamante.train)

Residuals:
     Min       1Q   Median       3Q      Max 
-17852.6   -651.8   -147.2    429.2  12563.2 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)    
(Intercept) 15095.492    470.576  32.079   <2e-16 ***
quilates    10875.145     65.724 165.467   <2e-16 ***
talla          56.543      6.730   8.402   <2e-16 ***
color        -262.839      4.116 -63.855   <2e-16 ***
claridad      289.123      4.002  72.246   <2e-16 ***
profundidad  -147.445      5.684 -25.939   <2e-16 ***
tabla         -88.087      3.246 -27.136   <2e-16 ***
x           -2197.414    105.230 -20.882   <2e-16 ***
y            1138.733    104.789  10.867   <2e-16 ***
z             -52.381     41.370  -1.266    0.205    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 1344 on 40445 degrees of freedom
Multiple R-squared:  0.8865,	Adjusted R-squared:  0.8864 
F-statistic: 3.509e+04 on 9 and 40445 DF,  p-val

Podemos interpretar este resultado como que todos los parametros del modelo de regresion multiple, incluyendo el intercepto, son altamente significativos en el modelo, a excepción de la variable 'z' que no es significativa o no aporta ninguna información que permita decidir sobre la variable 'income'.

**R^2 ajustado : 0.8864**

In [12]:
modelo_step<-step(modelo_simple,direction="both",trace=0)
summary(modelo_step)


Call:
lm(formula = precio ~ quilates + talla + color + claridad + profundidad + 
    tabla + x + y, data = diamante.train)

Residuals:
     Min       1Q   Median       3Q      Max 
-17863.6   -651.8   -147.1    429.2  12563.8 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)    
(Intercept) 15285.296    446.065  34.267   <2e-16 ***
quilates    10875.474     65.724 165.472   <2e-16 ***
talla          56.486      6.730   8.394   <2e-16 ***
color        -262.816      4.116 -63.849   <2e-16 ***
claridad      289.102      4.002  72.241   <2e-16 ***
profundidad  -150.547      5.129 -29.351   <2e-16 ***
tabla         -88.056      3.246 -27.127   <2e-16 ***
x           -2216.308    104.167 -21.276   <2e-16 ***
y            1125.262    104.248  10.794   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 1344 on 40446 degrees of freedom
Multiple R-squared:  0.8865,	Adjusted R-squared:  0.8864 
F-statistic: 3.947e+04 on 8 and 4

**R^2 ajustado : 0.8864**

### EVALUANDO EL MODELO STEP

In [60]:
predicciones <- predict(modelo_step, newdata = diamante.test)
comparacion <- data.frame(Real = diamante.test$precio, Predicción = predicciones)
#head(comparacion,10)

In [13]:
if (!require(Metrics)) install.packages("Metrics")
library(Metrics)

Loading required package: Metrics

Warning message in library(package, lib.loc = lib.loc, character.only = TRUE, logical.return = TRUE, :
“there is no package called ‘Metrics’”
Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



In [61]:
# Calcular el MSE, RMSE y MAE
mse <- mse(comparacion$Real, comparacion$Predicción)
rmse <- rmse(comparacion$Real, comparacion$Predicción)
mae <- mae(comparacion$Real, comparacion$Predicción)

print(paste("MSE:", mse))
print(paste("RMSE:", rmse))
print(paste("MAE:", mae))

[1] "MSE: 2199723.88673233"
[1] "RMSE: 1483.14661673495"
[1] "MAE: 877.150890761825"


Podemos ver que el error RMSE el cuál tiene las mismas unidades del precio del diamante es un error MUY grande de 1483.14 . Por lo que el modelo no esta haciendo un buen trabajo en predecir el precio aproximado de los diamantes.

- El modelo de regresion lineal tienen los supuestos de independencia entre las variables explicativas.
- El modelo de regresion lineal tiene el supuesto de normalidad de los residuales
- El modelo de regresion lineal tiene el supuesto de Homocedasticidad

Seguramente los supuestos no estan cumpliendose y esto afecta los resultados de este modelo. Por lo que es necesario regularización.

**"RMSE: 1483.14661673495"** y **R^2 ajustado : 0.8864**

### **MODELO ELASTIC-NET**

In [21]:
if (!require(glmnet)) install.packages("glmnet")
library(glmnet)

Loading required package: glmnet

Warning message in library(package, lib.loc = lib.loc, character.only = TRUE, logical.return = TRUE, :
“there is no package called ‘glmnet’”
Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

also installing the dependencies ‘iterators’, ‘foreach’, ‘shape’, ‘RcppEigen’


Loading required package: Matrix

Loaded glmnet 4.1-8



In [31]:
# Para usar la librería glmnet es necesario crear matrices separadas para los predictores y la variable pendiente
predic.train <- as.matrix(diamante.train[, -which(names(diamante.train) == "precio")])
precio.train <- as.matrix(diamante.train$precio)

predic.test <- as.matrix(diamante.test[, -which(names(diamante.test) == "precio")])
precio.test <- as.matrix(diamante.test$precio)

### VALIDACION CRUZADA : λ : ELASTIC-NET

In [23]:
# aqui hago redes elasticas de 0.1 a 0.9
for (i in 1:10){
  assign(paste("found", i, sep=""), cv.glmnet(predic.train, precio.train, nfolds=5,
                                              alpha=i/10,))}

In [24]:
# Extraer los valores mínimos de cvm para cada modelo
min_cvm_values <- sapply(1:10, function(i) min(get(paste("found", i, sep=""))$cvm))
# Crear una tabla con los resultados
results_table <- data.frame(
  Model = paste("found", 1:10, sep=""),
  Alpha = seq(0.1, 1.0, by=0.1),
  Min_CVM = min_cvm_values
)

# Mostrar la tabla
print(results_table)

     Model Alpha Min_CVM
1   found1   0.1 1815574
2   found2   0.2 1814493
3   found3   0.3 1814710
4   found4   0.4 1815076
5   found5   0.5 1814774
6   found6   0.6 1813743
7   found7   0.7 1814574
8   found8   0.8 1814901
9   found9   0.9 1815322
10 found10   1.0 1814474


Podemos observar como $λ=0.6$ arroja el valor minimo de 1'813.743

### MODELO : ELASTIC NET

In [68]:
elastic6<-glmnet(predic.train,precio.train,alpha = 0.6)
coef(elastic6,s=found6$lambda.min)

10 x 1 sparse Matrix of class "dgCMatrix"
                     s1
(Intercept) 15056.60222
quilates    10792.28971
talla          60.38648
color        -262.05063
claridad      289.95333
profundidad  -146.80484
tabla         -90.24704
x           -1549.71204
y             528.53067
z             -63.46576

## EVALUACION

In [74]:
predicciones <-predict.glmnet(elastic6, predic.test, s=elastic6$lambda.min)

In [75]:
#calcular el RMSE
erroreselastic <- sqrt(mean((predicciones - precio.test)^2))
erroreselastic

[1] 1766.857

### ECUACION FINAL CON LOS PARAMETROS OPTIMIZADOS:

Precio= 15056.60222 + 10792.28971⋅**quilates** + 60.38648⋅**talla** − 262.05063⋅**color** + 289.95333⋅**claridad** − 146.80484⋅**profundidad** − 90.24704⋅**tabla** − 1549.71204⋅**x** + 528.53067⋅**y** − 63.46576⋅**z**

# ¿Tiene un mejor desempeño que el modelo sin regularización?
No.

Segun el modelo step el RMSE era de 1483.14661673495. Por otro lado, despues de hacer regularizacion el nuevo modelo tiene un RMSE de 1766.856, el cual es mucho mayor que el obtenido inicialmente.


Indicando que la regularizacion ELASTIC NET con el $λ$ más optimo no está ayudando a disminuir el RMSE en la prediccion de los precios de los diamantes en el dataset de prueba.